src/Shared/DiagnosticSourceInstrumentation/PropertyFetcher.cs

```diff
@@ -77,61 +68,119 @@ public bool TryFetch(object obj, out T value, bool skipObjNullCheck = false)
-                property = type.GetProperty(propertyName);
-            }
+            var property = type.DeclaredProperties.FirstOrDefault(p => string.Equals(p.Name, propertyName, StringComparison.OrdinalIgnoreCase)) ?? type.GetProperty(propertyName);
+            return CreateFetcherForProperty(property!)!;
 
-            return CreateFetcherForProperty(property);
-
-            static PropertyFetch CreateFetcherForProperty(PropertyInfo propertyInfo)
+            static PropertyFetch? CreateFetcherForProperty(PropertyInfo propertyInfo)
             {
                 if (propertyInfo == null || !typeof(T).IsAssignableFrom(propertyInfo.PropertyType))
                 {
                     // returns null and wait for a valid payload to arrive.
                     return null;
                 }
 
-                var typedPropertyFetcher = typeof(TypedPropertyFetch<,>);
-                var instantiatedTypedPropertyFetcher = typedPropertyFetcher.MakeGenericType(
-                    typeof(T), propertyInfo.DeclaringType, propertyInfo.PropertyType);
-                return (PropertyFetch)Activator.CreateInstance(instantiatedTypedPropertyFetcher, propertyInfo);
+                var declaringType = propertyInfo.DeclaringType;
+                if (declaringType!.IsValueType)
+                {
+                    throw new NotSupportedException(
+                        $"Type: {declaringType.FullName} is a value type. PropertyFetcher can only operate on reference payload types.");
+                }
+
+                if (declaringType == typeof(object))
+                {
+                    // This is only necessary on .NET 7.
+                    // It was used to force the AOT compiler to create an instantiation of the method with a reference type.
+                    // The code for that instantiation can then be reused at runtime to create instantiation over any other reference.
+                    // TODO: when adding net8.0 as the project target:
+                    // This is not needed in.NET 8, because the compiler is improved and call into MakeGenericMethod will be AOT-compatible.
+                    return CreateInstantiated<object>(propertyInfo);
+                }
+                else
+                {
+                    return DynamicInstantiationHelper(declaringType, propertyInfo);
+                }
+
+                // Separated as local function to be able to target the suppression to just this call
+                // IL3050 was generated here because of the call to MakeGenericType, which is problematic in AOT if one of the type parameters is a value type;
+                // because the compiler might need to generate code specific to that type.
+                // If the type parameter is reference type, there will be no problem; because the generated code can be shared among all reference type instantiations.
+#if NET6_0_OR_GREATER
+                [RequiresUnreferencedCode(TrimCompatibilityMessage)]
```

Comment:
Why do we need `RequiresUnreferencedCode` here? It should not be needed. Only the AOT suppression should be needed here.